File: localize.cpp
/*
 * localize.cpp — GPS-free visual localization (C++ port of localize.py)
 *
 * Reads index/sift_index_z18.bin (produced by export_index.py).
 *
 * Build:
 *   mkdir build && cd build && cmake .. && make -j4
 *
 * Usage:
 *   ./localize photo.jpg
 *   ./localize photo.jpg --show
 */

#include <opencv2/opencv.hpp>
#include <opencv2/features2d.hpp>
#include "cuda_bf_matcher.h"
#include "popsift_sift.h"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>

namespace fs = std::filesystem;
#include <string>
#include <vector>

// ── config (mirrors config.py) ────────────────────────────────────────────────

static bool        g_gpu_sift    = false;   // set true if PopSIFT init succeeds
static const int   TILE_SIZE     = 256;
static const float MATCH_RATIO   = 0.75f;
static const int   MIN_INLIERS   = 15;
static const int   TOP_CANDIDATES = 5;
static const char* TILE_DIR      = "tiles";
static const char* INDEX_DIR     = "index";

// ── coordinate math (mirrors tile_utils.py) ───────────────────────────────────

static std::pair<double,double> pixel_to_latlon(
    double px, double py, int tile_x, int tile_y, int zoom)
{
    int    n      = 1 << zoom;
    double wx     = tile_x + px / TILE_SIZE;
    double wy     = tile_y + py / TILE_SIZE;
    double lon    = wx / n * 360.0 - 180.0;
    double lat    = std::atan(std::sinh(M_PI * (1.0 - 2.0 * wy / n))) * 180.0 / M_PI;
    return {lat, lon};
}

// ── index ─────────────────────────────────────────────────────────────────────

struct TileEntry {
    int    tile_x, tile_y, zoom;
    cv::Mat descs;          // rows×128, CV_32F
};

static std::vector<TileEntry> load_index(const std::string& path)
{
    std::ifstream f(path, std::ios::binary);
    if (!f) throw std::runtime_error("Cannot open index: " + path);

    int32_t num_tiles = 0;
    f.read(reinterpret_cast<char*>(&num_tiles), sizeof(int32_t));

    std::vector<TileEntry> idx;
    idx.reserve(num_tiles);

    for (int i = 0; i < num_tiles; ++i) {
        TileEntry e;
        int32_t n_descs = 0;
        f.read(reinterpret_cast<char*>(&e.tile_x),  sizeof(int32_t));
        f.read(reinterpret_cast<char*>(&e.tile_y),  sizeof(int32_t));
        f.read(reinterpret_cast<char*>(&e.zoom),    sizeof(int32_t));
        f.read(reinterpret_cast<char*>(&n_descs),   sizeof(int32_t));
        e.descs = cv::Mat(n_descs, 128, CV_32F);
        f.read(reinterpret_cast<char*>(e.descs.data),
               (std::streamsize)n_descs * 128 * sizeof(float));
        idx.push_back(std::move(e));
    }
    return idx;
}

// ── Phase 1: FLANN — build once, query per frame ─────────────────────────────

struct FlannIndex {
    cv::Ptr<cv::FlannBasedMatcher> matcher;
    std::vector<int>               offsets;   // tile boundary offsets into all_descs
    int                            total_kp;
};

// Called once at startup — expensive (KD-tree build over all tile descriptors).
static FlannIndex build_flann(const std::vector<TileEntry>& index)
{
    FlannIndex fi;
    fi.offsets.resize(index.size() + 1, 0);
    for (size_t i = 0; i < index.size(); ++i)
        fi.offsets[i + 1] = fi.offsets[i] + index[i].descs.rows;

    fi.total_kp = fi.offsets.back();
    cv::Mat all_descs(fi.total_kp, 128, CV_32F);
    for (size_t i = 0; i < index.size(); ++i)
        index[i].descs.copyTo(all_descs.rowRange(fi.offsets[i], fi.offsets[i+1]));

    std::cout << "Total train kp: " << fi.total_kp << "  — building FLANN …" << std::flush;
    fi.matcher = cv::makePtr<cv::FlannBasedMatcher>(
        cv::makePtr<cv::flann::KDTreeIndexParams>(5),
        cv::makePtr<cv::flann::SearchParams>(50));
    fi.matcher->add(all_descs);
    fi.matcher->train();
    std::cout << " done.\n";
    return fi;
}

// Called per frame — fast (KD-tree already built).
static std::vector<std::pair<int,int>> flann_vote(
    const cv::Mat& descs_q,
    const FlannIndex& fi,
    size_t n_tiles)
{
    std::vector<std::vector<cv::DMatch>> raw;
    fi.matcher->knnMatch(descs_q, raw, 2);

    std::vector<int> votes(n_tiles, 0);
    int good_count = 0;
    for (auto& pair : raw) {
        if (pair.size() < 2) continue;
        if (pair[0].distance < MATCH_RATIO * pair[1].distance) {
            ++good_count;
            int train_idx = pair[0].trainIdx;
            int tile_i = (int)(std::upper_bound(
                fi.offsets.begin(), fi.offsets.end(), train_idx)
                - fi.offsets.begin()) - 1;
            if (tile_i >= 0 && tile_i < (int)n_tiles)
                votes[tile_i]++;
        }
    }
    std::cout << "Good matches  : " << good_count << " (after ratio test)\n";

    std::vector<std::pair<int,int>> ranked;
    for (size_t i = 0; i < n_tiles; ++i)
        if (votes[i] > 0) ranked.push_back({(int)i, (int)votes[i]});
    std::sort(ranked.begin(), ranked.end(),
              [](auto& a, auto& b){ return a.second > b.second; });
    return ranked;
}

// ── SIFT wrapper: GPU (PopSIFT) with CPU (cv::SIFT) fallback ─────────────────

static void detect_compute(const cv::Mat& gray, std::vector<cv::KeyPoint>& kps,
                            cv::Mat& descs, int max_kp)
{
    if (g_gpu_sift) {
        popsift_detect_compute(gray, kps, descs, max_kp);
    } else {
        auto sift = cv::SIFT::create(max_kp > 0 ? max_kp : 2000);
        sift->detectAndCompute(gray, cv::noArray(), kps, descs);
    }
}

// ── Phase 2: stitched direct match ────────────────────────────────────────────

static cv::Mat stitch_tiles(int cx, int cy, int zoom, const char* tile_dir, int radius = 1)
{
    int size = 2 * radius + 1;
    cv::Mat canvas(size * TILE_SIZE, size * TILE_SIZE, CV_8UC3, cv::Scalar(0, 0, 0));
    for (int dx = 0; dx < size; ++dx) {
        for (int dy = 0; dy < size; ++dy) {
            int tx = cx - radius + dx;
            int ty = cy - radius + dy;
            std::string path = std::string(tile_dir) + "/" +
                               std::to_string(zoom) + "/" +
                               std::to_string(tx) + "/" +
                               std::to_string(ty) + ".png";
            cv::Mat tile = cv::imread(path);
            if (!tile.empty())
                tile.copyTo(canvas(cv::Rect(dx*TILE_SIZE, dy*TILE_SIZE, TILE_SIZE, TILE_SIZE)));
        }
    }
    return canvas;
}

struct MatchResult {
    double lat, lon;
    int    inliers;
    cv::Mat H, composite;
};

static std::optional<MatchResult> direct_match(
    const cv::Mat& query_gray,
    const std::vector<cv::KeyPoint>& kps_q,
    const cv::Mat& descs_q,
    int cx, int cy, int zoom,
    const char* tile_dir,
    int radius = 1, int min_inliers = 4)
{
    using clk = std::chrono::high_resolution_clock;
    auto t_sift0 = clk::now();

    cv::Mat composite = stitch_tiles(cx, cy, zoom, tile_dir, radius);
    cv::Mat comp_gray;
    cv::cvtColor(composite, comp_gray, cv::COLOR_BGR2GRAY);

    // Extract features — GPU (PopSIFT) if available, else CPU (cv::SIFT).
    // Both composite and query use the same extractor for descriptor compatibility.
    std::vector<cv::KeyPoint> kps_c, kps_q2;
    cv::Mat descs_c, descs_q2;
    detect_compute(comp_gray,  kps_c,  descs_c,  3000);
    detect_compute(query_gray, kps_q2, descs_q2, 2000);
    long sift_ms = std::chrono::duration_cast<std::chrono::milliseconds>(clk::now()-t_sift0).count();

    if (descs_c.empty()  || (int)kps_c.size()  < min_inliers)
        return std::nullopt;
    if (descs_q2.empty() || (int)kps_q2.size() < min_inliers)
        return std::nullopt;

    // Match descriptors — CUDA BFMatcher if GPU available, else OpenCV BFMatcher
    auto t_match0 = clk::now();
    std::vector<std::vector<cv::DMatch>> raw;
    if (g_gpu_sift) {
        cuda_knn2_match(descs_q2, descs_c, raw);
    } else {
        auto bf = cv::BFMatcher::create(cv::NORM_L2);
        bf->knnMatch(descs_q2, descs_c, raw, 2);
    }
    long match_ms = std::chrono::duration_cast<std::chrono::milliseconds>(clk::now()-t_match0).count();
    std::cout << "  [Phase2 SIFT=" << sift_ms << "ms  Match=" << match_ms << "ms";

    std::vector<cv::DMatch> good;
    for (auto& pair : raw)
        if (pair.size() >= 2 && pair[0].distance < MATCH_RATIO * pair[1].distance)
            good.push_back(pair[0]);

    if ((int)good.size() < min_inliers) {
        std::cout << "  ratio=" << good.size() << "/" << raw.size() << "]\n";
        return std::nullopt;
    }

    auto t_ransac0 = clk::now();
    std::vector<cv::Point2f> src_pts, dst_pts;
    for (auto& m : good) {
        src_pts.push_back(kps_q2[m.queryIdx].pt);
        dst_pts.push_back(kps_c[m.trainIdx].pt);
    }

    cv::Mat mask;
    cv::Mat H = cv::findHomography(src_pts, dst_pts, cv::RANSAC, 5.0, mask);
    long ransac_ms = std::chrono::duration_cast<std::chrono::milliseconds>(clk::now()-t_ransac0).count();
    std::cout << "  RANSAC=" << ransac_ms << "ms]\n";
    if (H.empty()) return std::nullopt;

    int inliers = cv::countNonZero(mask);
    if (inliers < min_inliers) {
        std::cout << "  [inliers=" << inliers << " < " << min_inliers << ", rejected]\n";
        return std::nullopt;
    }

    // Project query centre into composite pixel space
    float qw = (float)query_gray.cols, qh = (float)query_gray.rows;
    std::vector<cv::Point2f> ctr_in  = {{qw / 2, qh / 2}};
    std::vector<cv::Point2f> ctr_out;
    cv::perspectiveTransform(ctr_in, ctr_out, H);
    float cpx = ctr_out[0].x, cpy = ctr_out[0].y;

    int comp_size = (2 * radius + 1) * TILE_SIZE;
    int margin    = TILE_SIZE;
    if (cpx < -margin || cpx > comp_size + margin ||
        cpy < -margin || cpy > comp_size + margin)
        return std::nullopt;

    int   tile_dx = (int)(cpx / TILE_SIZE);
    int   tile_dy = (int)(cpy / TILE_SIZE);
    float local_px = cpx - tile_dx * TILE_SIZE;
    float local_py = cpy - tile_dy * TILE_SIZE;

    int x_orig = cx - radius;
    int y_orig = cy - radius;
    auto [lat, lon] = pixel_to_latlon(local_px, local_py,
                                      x_orig + tile_dx, y_orig + tile_dy, zoom);
    return MatchResult{lat, lon, inliers, H, composite};
}

// ── per-image localization ────────────────────────────────────────────────────

static void localize_one(const std::string& image_path, bool show,
                         const std::vector<TileEntry>& index,
                         const FlannIndex& fi,
                         int zoom_level)
{
    auto t0 = std::chrono::high_resolution_clock::now();
    auto elapsed_ms = [&t0]() {
        return std::chrono::duration_cast<std::chrono::milliseconds>(
            std::chrono::high_resolution_clock::now() - t0).count();
    };

    std::cout << "\n" << std::string(55, '-') << "\n";
    std::cout << "Image: " << image_path << "\n";

    cv::Mat query = cv::imread(image_path);
    if (query.empty()) {
        std::cerr << "  ERROR: cannot read image\n";
        return;
    }
    cv::Mat query_gray;
    cv::cvtColor(query, query_gray, cv::COLOR_BGR2GRAY);
    std::cout << "  Size    : " << query.cols << "×" << query.rows << " px\n";

    auto sift_q = cv::SIFT::create(2000);
    std::vector<cv::KeyPoint> kps_q;
    cv::Mat descs_q;
    sift_q->detectAndCompute(query_gray, cv::noArray(), kps_q, descs_q);
    std::cout << "  Features: " << kps_q.size() << " keypoints\n";

    // Phase 1 — coarse vote (FLANN already built)
    auto ranked = flann_vote(descs_q, fi, index.size());

    std::cout << "  Top candidates:\n";
    for (int i = 0; i < std::min((int)ranked.size(), TOP_CANDIDATES); ++i) {
        auto [ti, votes] = ranked[i];
        std::cout << "    tile (" << index[ti].tile_x << "," << index[ti].tile_y
                  << ")  votes=" << votes << "\n";
    }

    // Phase 2 — stitched direct match
    std::optional<MatchResult> result;
    int result_cx = -1, result_cy = -1;

    for (int i = 0; i < std::min((int)ranked.size(), TOP_CANDIDATES); ++i) {
        int ti = ranked[i].first;
        int cx = index[ti].tile_x;
        int cy = index[ti].tile_y;
        result = direct_match(query_gray, kps_q, descs_q,
                              cx, cy, zoom_level, TILE_DIR, 1, MIN_INLIERS);
        if (result) { result_cx = cx; result_cy = cy; break; }
    }

    long total_ms = elapsed_ms();

    if (!result) {
        std::cout << "  RESULT  : Could not determine location.\n";
        std::cout << "  Time    : " << std::fixed << std::setprecision(2)
                  << total_ms / 1000.0 << " s\n";
        return;
    }

    std::cout << std::fixed << std::setprecision(7);
    std::cout << "  GPS     : " << result->lat << ",  " << result->lon << "\n";
    std::cout << "  Inliers : " << result->inliers
              << "  |  tile (" << result_cx << "," << result_cy << ") z" << zoom_level << "\n";
    std::cout << "  Maps    : https://maps.google.com/?q="
              << result->lat << "," << result->lon << "\n";
    std::cout << std::setprecision(2) << "  Time    : " << total_ms / 1000.0 << " s\n";

    if (show) {
        float qw = (float)query_gray.cols, qh = (float)query_gray.rows;
        std::vector<cv::Point2f> corners_in = {{0,0},{qw,0},{qw,qh},{0,qh}};
        std::vector<cv::Point2f> corners_out;
        cv::perspectiveTransform(corners_in, corners_out, result->H);

        cv::Mat vis = result->composite.clone();
        std::vector<cv::Point> pts;
        for (auto& p : corners_out) pts.push_back({(int)p.x, (int)p.y});
        cv::polylines(vis, pts, true, {0, 255, 0}, 2);

        std::vector<cv::Point2f> ctr_in  = {{qw / 2, qh / 2}};
        std::vector<cv::Point2f> ctr_out;
        cv::perspectiveTransform(ctr_in, ctr_out, result->H);
        cv::drawMarker(vis, {(int)ctr_out[0].x, (int)ctr_out[0].y},
                       {0, 0, 255}, cv::MARKER_CROSS, 20, 2);

        // Save per-image result files
        fs::path p(image_path);
        std::string stem = p.stem().string();
        std::string tile_out  = "result_tile_"  + stem + ".png";
        std::string query_out = "result_query_" + stem + ".png";
        cv::imwrite(tile_out,  vis);
        cv::imwrite(query_out, query);
        std::cout << "  Saved   : " << tile_out << "  " << query_out << "\n";
    }
}

// ── main ─────────────────────────────────────────────────────────────────────

int main(int argc, char* argv[])
{
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " image1 [image2 ...] [--show] [--zoom N]\n";
        return 1;
    }

    // Split args into image paths and flags
    std::vector<std::string> images;
    bool show = false;
    int zoom_level = 18;
    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "--show") show = true;
        else if (std::string(argv[i]) == "--zoom" && i + 1 < argc)
            zoom_level = std::stoi(argv[++i]);
        else images.push_back(argv[i]);
    }
    if (images.empty()) {
        std::cerr << "No image files specified.\n";
        return 1;
    }

    // ── one-time startup (done before takeoff) ───────────────────────────────
    std::string index_path = std::string(INDEX_DIR) + "/sift_index_z" +
                             std::to_string(zoom_level) + ".bin";
    std::cout << "Loading index from " << index_path << " …\n";

    auto t_startup = std::chrono::high_resolution_clock::now();

    // Check available RAM before attempting GPU init.
    // PopSIFT calls abort() on CUDA OOM (not catchable), so we pre-screen.
    auto avail_mb = []() -> long {
        std::ifstream f("/proc/meminfo");
        std::string key; long val; std::string unit;
        while (f >> key >> val >> unit) {
            if (key == "MemAvailable:") return val / 1024;
        }
        return 0;
    };

    long mem_mb = avail_mb();
    std::cout << "Initialising GPU (PopSift + cuBLAS) … (avail=" << mem_mb << " MB) " << std::flush;
    if (mem_mb >= 512) {
        try {
            init_popsift();
            { cv::Mat tiny(4,128,CV_32F,cv::Scalar(0));
              std::vector<std::vector<cv::DMatch>> dummy;
              cuda_knn2_match(tiny, tiny, dummy); }
            g_gpu_sift = true;
            std::cout << "done (GPU).\n";
        } catch (const std::exception& e) {
            std::cout << "FAILED (" << e.what() << ")\n";
            std::cout << "Falling back to CPU SIFT.\n";
            g_gpu_sift = false;
        }
    } else {
        std::cout << "skipped — low RAM, using CPU SIFT.\n";
        g_gpu_sift = false;
    }

    auto index = load_index(index_path);
    long load_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - t_startup).count();
    std::cout << "Tile index    : " << index.size()
              << " tiles  (loaded in " << load_ms << " ms)\n";

    auto fi = build_flann(index);
    long startup_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - t_startup).count();
    std::cout << "Startup total : " << std::fixed << std::setprecision(2)
              << startup_ms / 1000.0 << " s  (index load + FLANN build)\n";

    // ── per-frame loop (each call is one drone camera frame) ─────────────────
    auto t_all = std::chrono::high_resolution_clock::now();
    for (auto& img_path : images)
        localize_one(img_path, show, index, fi, zoom_level);

    long total_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::high_resolution_clock::now() - t_all).count();
    std::cout << "\n" << std::string(55, '=') << "\n";
    std::cout << "  " << images.size() << " images processed in "
              << std::fixed << std::setprecision(2) << total_ms / 1000.0 << " s"
              << "  (index load excluded)\n";
    std::cout << std::string(55, '=') << "\n";
    shutdown_popsift();
    return 0;
}

File: localize.py
"""
Identify the GPS location of a query photo by matching it against the tile index.

Algorithm
---------
Phase 1 — coarse (FLANN vote):
  Extract SIFT from query; match against the full tile descriptor pool;
  vote to find which tile cluster wins.

Phase 2 — fine (stitched direct match):
  Stitch the winning tile + its 8 neighbours into a 3×3 composite;
  run SIFT+BFMatcher directly on the composite — more matching surface,
  no tile-boundary splits, far more inliers than Phase 1 alone.

Usage:
    python localize.py photo.jpg
    python localize.py photo.jpg --zoom 17 --show
"""

import argparse
import bisect
import pickle
from collections import defaultdict
from pathlib import Path

import cv2
import numpy as np

from config import (ZOOM_LEVEL, INDEX_DIR, TILE_DIR,
                    MATCH_RATIO, MIN_INLIERS, TOP_CANDIDATES, TILE_SIZE)
from tile_utils import pixel_to_latlon, tile_bounds


# ── index helpers ─────────────────────────────────────────────────────────────

def load_index(index_dir, zoom):
    path = Path(index_dir) / f'sift_index_z{zoom}.pkl'
    if not path.exists():
        raise FileNotFoundError(
            f'Index not found: {path}\nRun build_index.py first.')
    print(f'Loading index from {path} …')
    with open(path, 'rb') as f:
        return pickle.load(f)


def kp_from_tuple(t):
    return cv2.KeyPoint(x=t[0][0], y=t[0][1], size=t[1], angle=t[2],
                        response=t[3], octave=t[4], class_id=int(t[5]))


# ── Phase 1: FLANN vote ───────────────────────────────────────────────────────

def flann_vote(descs_q, index):
    """
    Match query descriptors against all tile descriptors via FLANN.
    Returns ranked list of (tile_idx, [matches]) sorted by vote count.
    """
    tile_offsets = [0]
    for entry in index:
        tile_offsets.append(tile_offsets[-1] + len(entry['descs']))
    all_descs = np.vstack([e['descs'] for e in index]).astype(np.float32)

    print(f'Total train kp: {len(all_descs):,}  — building FLANN …')
    index_params = dict(algorithm=1, trees=5)   # FLANN_INDEX_KDTREE
    search_params = dict(checks=50)
    flann = cv2.FlannBasedMatcher(index_params, search_params)
    flann.add([all_descs])
    flann.train()

    raw = flann.knnMatch(descs_q.astype(np.float32), k=2)
    good = [m for m, n in raw if len((m, n)) == 2 and
            m.distance < MATCH_RATIO * n.distance]

    print(f'Good matches  : {len(good)} (after ratio test)')

    votes = defaultdict(list)
    for m in good:
        tile_idx = bisect.bisect_right(tile_offsets, m.trainIdx) - 1
        votes[tile_idx].append(m)

    return sorted(votes.items(), key=lambda kv: len(kv[1]), reverse=True)


# ── Phase 2: stitched direct match ────────────────────────────────────────────

def stitch_tiles(cx, cy, zoom, tile_dir, radius=1):
    """
    Load a (2r+1)×(2r+1) grid of tiles centred on (cx, cy) and stitch
    them into one BGR image.  Returns (composite_bgr, x_origin, y_origin).
    """
    r = radius
    size = 2 * r + 1
    canvas = np.zeros((size * TILE_SIZE, size * TILE_SIZE, 3), dtype=np.uint8)
    for dx in range(size):
        for dy in range(size):
            tx, ty = cx - r + dx, cy - r + dy
            path = tile_dir / str(zoom) / str(tx) / f'{ty}.png'
            tile = cv2.imread(str(path))
            if tile is not None:
                canvas[dy * TILE_SIZE:(dy + 1) * TILE_SIZE,
                       dx * TILE_SIZE:(dx + 1) * TILE_SIZE] = tile
    return canvas, cx - r, cy - r


def direct_match(query_gray, kps_q, descs_q, cx, cy, zoom, tile_dir,
                 radius=1, min_inliers=4):
    """
    Stitch a (2*radius+1)² tile grid around (cx, cy), run SIFT+BFMatcher,
    RANSAC homography, and return (lat, lon, inliers) or None.
    """
    composite, x_orig, y_orig = stitch_tiles(cx, cy, zoom, tile_dir, radius)
    comp_gray = cv2.cvtColor(composite, cv2.COLOR_BGR2GRAY)

    sift = cv2.SIFT_create(nfeatures=3000)
    kps_c, descs_c = sift.detectAndCompute(comp_gray, None)
    if descs_c is None or len(kps_c) < min_inliers:
        return None

    bf = cv2.BFMatcher(cv2.NORM_L2)
    raw = bf.knnMatch(descs_q, descs_c, k=2)
    good = [m for m, n in raw if m.distance < MATCH_RATIO * n.distance]

    if len(good) < min_inliers:
        return None

    src_pts = np.float32([kps_q[m.queryIdx].pt for m in good]).reshape(-1, 1, 2)
    dst_pts = np.float32([kps_c[m.trainIdx].pt for m in good]).reshape(-1, 1, 2)

    H, mask = cv2.findHomography(src_pts, dst_pts, cv2.RANSAC, 5.0)
    if H is None:
        return None

    inliers = int(mask.ravel().sum())
    if inliers < min_inliers:
        return None

    # Reject degenerate homographies: project the four query corners and check
    # that the resulting quadrilateral has a reasonable area and aspect ratio.
    h, w = query_gray.shape
    corners = cv2.perspectiveTransform(
        np.float32([[[0,0],[w,0],[w,h],[0,h]]]), H)[0]
    area = cv2.contourArea(corners)
    if area < (w * h * 0.05) or area > (w * h * 50):
        return None
    # Check convexity (degenerate homographies produce concave quads)
    if not cv2.isContourConvex(np.int32(corners)):
        return None

    # Project query centre into composite pixel space
    h, w = query_gray.shape
    centre = cv2.perspectiveTransform(
        np.float32([[[w / 2, h / 2]]]), H)[0][0]
    cpx, cpy = float(centre[0]), float(centre[1])

    # Composite pixel → individual tile + local pixel → GPS
    comp_size = (2 * radius + 1) * TILE_SIZE
    margin = TILE_SIZE
    if not (-margin <= cpx <= comp_size + margin and
            -margin <= cpy <= comp_size + margin):
        return None

    tile_dx = int(cpx / TILE_SIZE)
    tile_dy = int(cpy / TILE_SIZE)
    local_px = cpx - tile_dx * TILE_SIZE
    local_py = cpy - tile_dy * TILE_SIZE
    lat, lon = pixel_to_latlon(local_px, local_py,
                               x_orig + tile_dx, y_orig + tile_dy, zoom)
    return lat, lon, inliers, H, composite


# ── main ─────────────────────────────────────────────────────────────────────

def localize(image_path, zoom=None, index_dir=None, tile_dir=None, show=False):
    zoom = zoom or ZOOM_LEVEL
    index_dir = index_dir or INDEX_DIR
    tile_dir = Path(tile_dir or TILE_DIR)

    # Load query
    query = cv2.imread(str(image_path))
    if query is None:
        raise FileNotFoundError(f'Cannot read image: {image_path}')
    query_gray = cv2.cvtColor(query, cv2.COLOR_BGR2GRAY)

    sift = cv2.SIFT_create(nfeatures=2000)
    kps_q, descs_q = sift.detectAndCompute(query_gray, None)
    if descs_q is None or len(kps_q) < 5:
        print('Too few features in query image.')
        return None
    print(f'Query image   : {image_path}  ({query.shape[1]}×{query.shape[0]} px)')
    print(f'Query features: {len(kps_q)} keypoints')

    # Phase 1 — coarse vote
    index = load_index(index_dir, zoom)
    print(f'Tile index    : {len(index)} tiles')
    ranked = flann_vote(descs_q, index)

    print('\nTop candidate tiles:')
    for idx, ms in ranked[:TOP_CANDIDATES]:
        e = index[idx]
        print(f'  tile ({e["tile_x"]},{e["tile_y"]})  votes={len(ms)}')

    # Phase 2 — stitched direct match on top candidates
    result = None
    for tile_idx, ms in ranked[:TOP_CANDIDATES]:
        entry = index[tile_idx]
        cx, cy = entry['tile_x'], entry['tile_y']
        print(f'\nDirect match on 3×3 stitched grid centred ({cx},{cy}) …')

        out = direct_match(query_gray, kps_q, descs_q,
                           cx, cy, zoom, tile_dir,
                           radius=1, min_inliers=MIN_INLIERS)
        if out is not None:
            lat, lon, inliers, H_comp, composite = out
            result = (lat, lon, inliers, cx, cy, H_comp, composite)
            break

    # Report
    print()
    if result is None:
        print('Could not determine location.')
        print('Suggestions:')
        print('  • Check that the photo location is within the downloaded tiles')
        print('  • Try a nadir (straight-down) photo with clear ground features')
        print('  • Increase SIFT_N_FEATURES in config.py and rebuild the index')
        return None

    lat, lon, inliers, cx, cy, H_comp, composite = result
    print('=' * 55)
    print(f'  GPS Location: {lat:.7f},  {lon:.7f}')
    print(f'  RANSAC inliers: {inliers}  |  centre tile ({cx},{cy}) z{zoom}')
    print(f'  Google Maps: https://maps.google.com/?q={lat:.7f},{lon:.7f}')
    print('=' * 55)

    if show:
        _save_result(query, query_gray, composite, H_comp, lat, lon)

    return lat, lon


# ── visualisation (saves to file, no Qt) ─────────────────────────────────────

def _save_result(query_img, query_gray, composite, H, lat, lon):
    h, w = query_gray.shape
    corners = np.float32([[0, 0], [w, 0], [w, h], [0, h]]).reshape(-1, 1, 2)
    projected = cv2.perspectiveTransform(corners, H)
    comp_vis = cv2.polylines(composite.copy(), [np.int32(projected)],
                             True, (0, 255, 0), 2)

    centre = cv2.perspectiveTransform(
        np.float32([[[w / 2, h / 2]]]), H)[0][0]
    cv2.drawMarker(comp_vis, (int(centre[0]), int(centre[1])),
                   (0, 0, 255), cv2.MARKER_CROSS, 20, 2)

    cv2.imwrite('result_tile.png', comp_vis)
    cv2.imwrite('result_query.png', query_img)
    print(f'Match saved → result_tile.png  (3×3 composite with footprint)')
    print(f'Query saved → result_query.png')


# ── entry point ───────────────────────────────────────────────────────────────

if __name__ == '__main__':
    parser = argparse.ArgumentParser(
        description='Identify GPS location of a photo using satellite map tiles')
    parser.add_argument('image', help='Path to the query photo')
    parser.add_argument('--zoom', type=int, default=None)
    parser.add_argument('--index-dir', default=None)
    parser.add_argument('--tile-dir', default=None)
    parser.add_argument('--show', action='store_true',
                        help='Save match visualisation to result_tile.png')
    args = parser.parse_args()

    localize(args.image, zoom=args.zoom, index_dir=args.index_dir,
             tile_dir=args.tile_dir, show=args.show)
